### PR TITLE
fix(docker): inspecting base images to find entrypoint

### PR DESCRIPTION
**Pull request by miki725**
- [x] Followed the steps in the contributor's guide: https://crashoverride.com/docs/other/contributing#filing-the-pull-request
- [x] PR title uses [semantic commit messages](https://nitayneeman.com/posts/understanding-semantic-commit-messages-using-git-and-angular/#fix)
- [x] Filled out the template to a useful degree

## Issue

any docker build with `ENTRYPOINT` defined in base image is not wrapped correctly

## Description

This recursively looks up entrypoints in all base images
from the dockerfile build target. If ENTRYPOINT/CMD/SHELL
is not found in the dockerfile, base image is inspected
to find that. This allows to correctly wrap all builds
as we know exactly how to wrap existing entrypoints.
This in particular will fix the issue for non-shell
ENTRYPOINTs such as for lambda wrappers which expect
CMD to be a python module argument.

In addition this fixes:

* adding docker statements within the target dockerfile section
  boundary. Previously it was always added at the end of Dockerfile
* it honors any --build-arg as passed in via CLI in order
  to eval env vars in the Dockerfile (needed for base image inspection)
* improves error message when docker build fails and fallback is used.
  now exception message will be showed.
* base image config is queried directly from the registry
  via buildx imagetools. This allows to avoid any local docker
  state changes. For example `docker pull alpine --platform=linux/arm64`
  will overwrite local `alpine` image to use arm which can impact
  future builds.
* switching all tests to use trace log level by default.
  this actually found a bug in multi-platform builds as --load
  was not being used by default and so full image metadata
  was not inspected

## Testing

```
make tests args="test_docker.py"
```

### CODE REVIEW

File: src/commands/cmd_docker.nim [hunk 1]
```diff
@@ -25,10 +25,21 @@
...
+proc addInstructions(ctx: DockerInvocation, content: string): string =
+  ## add instructions in correct location in Dockerfile
+  ## this accounts for section boundaries
+  let
+    section = ctx.getTargetDockerSection()
+    lines   = content.splitLines()
+    before  = lines[0 .. section.endLine]
+    after   = lines[section.endLine + 1 .. ^1]
+    updated = before & ctx.addedInstructions & after
+  return updated.join("\n")
```
[the remark this answers is not in the document]
Comment (miki725) [reply src/commands/cmd_docker.nim:41]: how would you slice it without `[]`. is there a diff function to use?

File: src/docker_base.nim [hunk 2]
```diff
@@ -251,6 +251,245 @@ proc getAllDockerContexts*(info: DockerInvocation): seq[string] =
...
+    envVars    = @[setEnv("DOCKER_BUILDKIT", "1")]
+    probeFile  = """
+FROM busybox
+ARG BUILDPLATFORM
+ARG TARGETPLATFORM
+RUN echo "{\"BUILDPLATFORM\": \"$BUILDPLATFORM\", \"TARGETPLATFORM\": \"$TARGETPLATFORM\"}" > /platforms.json
+CMD cat /platforms.json
+"""
+
+  var data: string
```
[the remark this answers is not in the document]
Comment (miki725) [reply src/docker_base.nim:278]: thanks. good to know. I saw this pattern a lot here so mostly started using it for consistency but Ill do that instead from now on. similar question but for tables? what is the expected pattern there?

https://nim-lang.org/docs/tables.html#initTable

> Starting from Nim v0.20, tables are initialized by default and it is not necessary to call this function explicitly.

as per above, would it still be the recommendation to initialize table with that function, even if not necessary:

```nim
let a = initTable[string, string]()
```

File: src/docker_base.nim [hunk 3]
```diff
@@ -251,6 +251,245 @@ proc getAllDockerContexts*(info: DockerInvocation): seq[string] =
...
+    if len(info.dfSections) == 0:
+      raise newException(ValueError, "there are no docker sections")
+    return info.dfSections[^1]
+  else:
+    if info.targetBuildStage notin info.dfSectionAliases:
+      raise newException(KeyError, info.targetBuildStage & ": is not found in Dockerfile")
+    return info.dfSectionAliases[info.targetBuildStage]
+
+proc getTargetEntrypoints*(info: DockerInvocation):
+  tuple[entrypoint: EntrypointInfo, cmd: CmdInfo, shell: ShellInfo] =
```
[the remark this answers is not in the document]
Comment (miki725) [reply src/docker_base.nim:450]: oh haha broke my own style. when the function declaration is over multiple lines I like to indent it differently so it doesnt have same indent level as rest of the function body. is there a standard way of doing that in nim? alternatively if function `()` is split into multiple lines the `)` is then on the root level of the file which I think is more readable and thats the approach `black` uses in python for long function definitions. not sure what is the best way in nim

```nim
proc foo(
  a: string,
  b: string,
): tuple[
  foo: string,
  bar: string,
] =
```

**Comment by miki725** [comment]
in terms of test coverage, I did a bunch of tests manually and I think test coverage has enough cases to cover what this PR is fixing. it had one explicit test for base image entrypoint inspection which was marked as failing which is now resolved so I think we should be ok there

**Review by viega — APPROVED** [review]

**Pull request merged**